sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+  for (auto attr_field : mio_attr_fields_) {
+    TestWriteAttrFailureConflict(attr_field, kDifPadctrlPadMio);
+  }
+
+  for (auto attr_field : dio_attr_fields_) {
+    TestWriteAttrFailureConflict(attr_field, kDifPadctrlPadDio);
+  }
+}
+
+TEST_F(DifPadctrlTest, EnableAttrFailureInvalidPad) {
+  DeviceInit();
+
+  EXPECT_EQ(
+      dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadMio,
+                               PADCTRL_PARAM_NMIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(
+      dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadDio,
+                               PADCTRL_PARAM_NDIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadMio, -1,
+                                     kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadDio, -1,
+                                     kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+}
+
+TEST_F(DifPadctrlTest, DisableAttrFailureInvalidPad) {
+  DeviceInit();
+
+  EXPECT_EQ(
+      dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadMio,
+                                PADCTRL_PARAM_NMIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(
+      dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadDio,
+                                PADCTRL_PARAM_NDIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadMio, -1,
+                                      kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadDio, -1,
+                                      kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+}
+
+TEST_F(DifPadctrlTest, WriteAttrFailureInvalidPad) {
+  DeviceInit();
+
+  EXPECT_EQ(
+      dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadMio,
+                              PADCTRL_PARAM_NMIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(
+      dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadDio,
+                              PADCTRL_PARAM_NDIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadMio, -1,
+                                    kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadDio, -1,
+                                    kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+}
+
+TEST_F(DifPadctrlTest, EnableAttrsFailureInvalidAttr) {
+  DeviceInit();
+
+  for (auto pad_kind : {kDifPadctrlPadMio, kDifPadctrlPadDio}) {
+    dif_padctrl_attr_t bad_attrs_1 = dif_padctrl_attr_t(
+        (1 << PADCTRL_PARAM_ATTRDW) | kDifPadctrlAttrOpenDrain);
+
```

Comment:
Here, again, remove this newline. `bad_attrs_1` is only used with the `EXPECT_EQ` below, so they aught to share a stanza.